src/Host/Client/Impl/Session/RSessionProvider.cs

```diff
@@ -249,20 +259,18 @@ public class RSessionProvider : IRSessionProvider {
             return true;
         }
 
-        private async Task SwitchBrokerAsync(IBrokerClient oldBroker, CancellationToken cancellationToken, ReentrancyToken reentrancyToken) {
-            var sessions = _sessions.Values.ToList();
-            if (sessions.Any()) {
-                await SwitchSessionsAsync(sessions, cancellationToken, reentrancyToken);
+        private async Task SwitchBrokerAsync(CancellationToken cancellationToken, ReentrancyToken reentrancyToken) {
+            var transactions = _sessions.Values.Select(s => s.StartSwitchingBroker()).ExcludeDefault().ToList();
```

Comment:
`IEnumerable<IRSession> GetSessions()` should have ToArray() since otherwise it is returning Values that can change.